### DISCUSSION TITLE
Does Pele include a turbulence forcing module?

### DISCUSSION THREAD

**Discussion by Alex-jian522** [AMReX-Combustion/PeleLMeX#447] [General]
Does pele include a turbulence forcing module? Or does it support adding user-defined source terms to the momentum equation? Thanks for your reply!

**Comment by baperry2** [comment]
You can add user-defined source terms through the `problem_modify_ext_sources` function. See [here](https://github.com/AMReX-Combustion/PeleLMeX/blob/55ef5ebb4ba1dc134275e635e501fc95fe42e96b/Exec/RegTests/EB_ODEQty/PeleLMeX_ProblemSpecificFunctions.cpp#L27) for an example where the source term is applied to scalars, but you could similarly do a momentum source.

**Reply by Alex-jian522** [reply]
Thanks very much! I will have a try.

**Reply by Alex-jian522** [reply]
I use the following linear forcing method to obtain quasi-steady turbulence. To do this, I need to obtain the average value of the whole domain at each moment, such as the product of the average density and the turbulent kinetic energy. How can I achieve this in PeleLMeX_ProblemSpecificFunctions.cpp? Is there any code that I can refer to? Thank you very much!
![turbulence_forcing](https://github.com/user-attachments/assets/9b090b6a-4817-42a9-9c94-827ea4256940)

**Reply by baperry2** [reply]
You can use `PeleLM::MFSum` to compute an integral over the domain in order to compute the domain-averaged quantities.

Unfortunately, the `problem_modify_ext_sources` operates on only a single level and you need data from all levels to compute the integral, so you'll have to modify the core source code to do what you want.
